README.md
# markit
[![GitHub Stable Release](https://img.shields.io/badge/Release-4.2.1-blue.svg)](https://github.com/BradleyA/markit/releases/tag/4.2.1)
![GitHub Release Date](https://img.shields.io/github/release-date/BradleyA/markit?color=blue)
[![GitHub Commits Since](https://img.shields.io/github/commits-since/BradleyA/markit/4.2.1?color=orange)](https://github.com/BradleyA/markit/commits/)
[![GitHub Last Commits](https://img.shields.io/github/last-commit/BradleyA/markit.svg)](https://github.com/BradleyA/markit/commits/)

[![Open GitHub Issue](https://img.shields.io/badge/Open-Incident-brightgreen.svg)](https://github.com/BradleyA/markit/issues/new/choose)
[![GitHub Open Issues](https://img.shields.io/github/issues/BradleyA/markit?color=purple)](https://github.com/BradleyA/markit/issues?q=is%3Aopen+is%3Aissue)
[![GitHub Closed Issues](https://img.shields.io/github/issues-closed/BradleyA/markit?color=purple)](https://github.com/BradleyA/markit/issues?q=is%3Aclosed+is%3Aissue)

[<img alt="GitHub Clones" src="https://img.shields.io/static/v1?label=Clones&message=554&color=blueviolet">](https://github.com/BradleyA/markit/blob/master/images/clone.table.md)
[<img alt="GitHub Views" src="https://img.shields.io/static/v1?label=Views&message=2631&color=blueviolet">](https://github.com/BradleyA/markit/blob/master/images/view.table.md)
[![GitHub Size](https://img.shields.io/github/repo-size/BradleyA/markit.svg)](https://github.com/BradleyA/markit/)
![Language Bash](https://img.shields.io/badge/%20Language-bash-blue.svg)
[![MIT License](http://img.shields.io/badge/License-MIT-blue.png)](LICENSE)

## Goal
Answer one question over the phone by looking at the code; ***What version are you running?**

#### If you like this repository, select in the upper-right corner, [![GitHub stars](https://img.shields.io/github/stars/BradleyA/markit.svg?style=social&label=Star&maxAge=2592000)](https://GitHub.com/BradleyA/markit/stargazers/), thank you.

## Table of content
<details>

- [Description](#Description)
- [Objectives](#Objectives)
- [Quick Start Tutorials](#Quick-Start-Tutorials)
- [Clone](#Clone)
- [Install latest markit command](#Install-latest-markit-command)
- [Usage: markit](#Usage-markit)
- [Output: markit](#Output-markit)
- [Usage: check-markit](#Usage-check-markit)
- [Output: check-markit](#Output-check-markit)
- [Usage: find-code.sh](#Usage-find-codesh)
- [Output: find-code.sh](#Output-find-codesh)
----
- [Contribute](#Contribute)
- [Author](#Author)
- [Tested OS](#Tested-OS)
- [Design Principles](#Design-Principles)
- [License](#License)

</details>

----
  
## Description
Bash script that adds Git Verion Control information into files and git add, git commit, git pushes them to GitHub, Bitbucket, and/or GitLab.   Most scripts in this repository support --help and --usage options.

**Markit** is a bash script that is a wrapper around these Git commands; git add ${FILE_NAME}, git commit -m "${FILE_MESSAGE}", git tag -a ${RELEASE_VERSION}, and git push --follow-tags. It automates this Git commit process and adds two comment lines of metadata about file changes to all tracked modified files in your local Git repository.  This metadata allows any person to answer the question; What version are you running? 

Markit supports [Semantic Versioning](https://semver.org/) (MAJOR.MINOR.PATCH) and alphanumeric versioning (v0.1.64-alpha).  Markit will display the current version before prompting you to enter the next version.  Markit includes the number of commits with the version in the file(s).  I find this helpful when commiting several different changes while working on the same PATCH.

If environment variable MARKITLOCAL is set to 1, markit automates this git commit process, but does not push to GitHub repository or Bitbucket repository or GitLab project.  I find this helpful when working without network access to a remote Git repository.  If using the bash shell, enter; 'export MARKITLOCAL=1' on the command line to set the MARKITLOCAL environment variable to '1'.  Use the command, 'unset MARKITLOCAL' to remove the exported information from the MARKITLOCAL environment variable.  You are on your own defining environment variables if you are using other shells.

The two comment lines of metadata include:

 * file_name, version.commits, date, time, time zone, repository URL, Git branch, user, hostname, previous version
 * file name(s) updated and a one line breif description about changes made to the file(s)
 
Markit currently supports files with these file extensions:

c(c), C++(cc|cpp|c++|cxx), C header(h|H|hpp), C++ header(hxx|Hxx|HXX), Go(go), HTML(html|htm), Java(java|class|jar), JavaScript(js), Kotlin(kt|kts), Makefile(mk|MK), Pascal(p|pp|paa), Perl(pl|PL|psh), PHP(php|php3|php4|ph3|ph4), PowerShell(ps1|msh), Python(py), R(R), Ruby(rb), Rust(rs|rlib), Scala(scala|sc), Shell(sh|bash|csh|tcsh|ksh), Terraform(tf|tfstate), Text(txt), XML(xml), YAML(yml|yaml)

Markit currently supports these files without file extensions: Dockerfile, Makefile.  There will be a prompt to add comment character(s) for unknown file names when needed and a prompt to add a description about the changes.

[Return to top](https://github.com/BradleyA/markit/blob/master/README.md#markit)

## Objectives
<details>
  
* Any person without knowledge of Git or GitHub or Bitbucket or GitLab or markit to be able to answer one question; 
  **What version are you running?**
* Supports [Semantic Versioning](https://semver.org/) (MAJOR.MINOR.PATCH) and alphanumeric versioning (v0.1.64-alpha)
* Only update modified files by inserting two comment lines with metadata
* Insert the lines of metadata after the first line in files
* Support GitHub
* Use git commit message as part of metadata
* Use git tag as part of metadata
* Include date and remote Git repository as part of metadata
* Include path with filename in remote Git repository as part of metadata
* Don't create any requirements for markit other than the version and those required by Git
* Support several file extensions for comment character
* Support Bitbucket and GitLab
* Support Git local repository without network access (without 'git push')
* Support push to multiple Git remote repositories at once (edit .git/config)
* I use alias release-notes='git log --graph --pretty=format:'\''%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cr) %C(bold blue)<%an>%Creset'\'' --abbrev-commit' to help build release notes

[Return to top](https://github.com/BradleyA/markit/blob/master/README.md#markit)

</details>

## Quick Start Tutorials

1) First make a copy of or fork this GitHub repository so you can modify the test files while learning how simple it is to work with markit.

* To fork a GitHub repository you need to be logged into your GitHub account.

* Go to repository **https://github.com/BradleyA/markit**

* Click the fork icon in the upper right

2) Second download or clone the forked copy of markit from your GitHub \<YOUR-USERNAME\>/markit to your Linux system.

* Enter the following command with the name of your repository,
    
      git clone https://github.com/<YOUR-USERNAME>/markit.git

* Change into the cloned directory on your Linux system.

      cd markit

3) Third make a change to the file, testfiles/sample.sh, then git tag, git commit, and git push the change to your local and remote GitHub repository.

* Make any change to testfiles/sample.sh.

      vi testfiles/sample.sh
    
* Run ./markit and enter the next version number when prompted and a brief description of the changes to testfiles/sample.sh.  You will be prompted for your GitHub password when Git pushes testfiles/sample.sh to your remote GitHub repository. 

    ./markit

 **That is it!**
 
Check if your local version of testfiles/sample.sh is the same version that is on your remote Git repository.

    ./check-markit -f ./testfiles/sample.sh

On GitHub go to your Forked markit repository settings page and check the box next to Issues. Then open the first GitHub issue on your forked markit Git repository, #1.

Modify test file, testfiles/sample.sh, by adding a blank line to learn how markit works.

    vi testfiles/sample.sh
    
    #!/bin/bash
    #   testfiles/sample.sh  3.197.368  2019-02-08T20:33:53.600661-06:00 (CST)  https://github.com/BradleyA/markit  uadmin  six-rpi3b.cptx86.com 3.196-1-g46f4c0d  
    #      test markit for sample.sh 
    #
    #   ./sample.sh
    ###
    
    echo "Hello World"

Run ./markit.  Enter the next version number when prompted and this brief description '[testing markit issue changes close #1](https://help.github.com/articles/closing-issues-using-keywords/)'.  You will be prompted for your GitHub password when git pushes testfiles/sample.sh to your remote forked markit Git repository.

    ./markit
    
The file testfiles/sample.sh has been updated with metadata, git tag, git commit, git push, and the issue #1 on GitHub has been [closed](https://help.github.com/articles/closing-issues-using-keywords/).


CHANGELOG.md
**Notes:** Use the following example to guide you to; How to download an earlier release.

    git clone https://github.com/BradleyA/<REPOSITORY>.git --branch 2.42

## Next Version Under Construction {Version  4.3.????}  <img id="Construction" src="images/construction-icon.gif" width="120">
#### Release Name  prerelease
#### Release Date  ????
* Brief description
* sub-repository
  * Brief description

### Features Added
* Updated feature||file
* Added the following test cases
    * hooks/bin/EXAMPLES/FVT-exit-code-error-0-001
* sub-repository
  * directory/command   git commit message
  * directory/command   git commit message closes #XX

### Features Changes
### Features Deprecations
### Issues
* close #X  directory/command   git commit message
* sub-repository
  * close #XX directory/command git commit message

### Misc
* backlog

# Latest Release --> Version  4.2.1.1114
#### Release Name  4.2.1
#### Release Date  2020-12-22T21:02:03.686774-06:00 (CST)
* Added test cases to be used with repository BradleyA/git-TEST-commit-automation 

### Features Added
* Updated .github/ISSUE_TEMPLATE/feature_request.md changed command example
* Updated .github/ISSUE_TEMPLATE/incident_report.md changed command example
* Updated .gitignore added hooks/ directory to .gitignore for git-TEST-commit-automation testing
* Updated README.md
* check-markit 
  * Added unique temporary directory names to support automation,  close #84
  * Added find-code.sh to check-markit SEE ALSO section
  * Check if file starts with '-', Invalid option
  * Change --file option to --filename option
  * Production standard 1.3.614 DEBUG variable
  * Production standard 2.3.614 Log format
  * Production standard 0.3.615 --help
  * Production standard 9.3.606 Parse CLI options and arguments
* find-code.sh markit
  * Change output color to help review of output
  * Redesign -c, -d, -s option code
  * Production standard 1.3.614 DEBUG variable
  * Production standard 2.3.614 Log format
  * Production standard 0.3.615 --help
  * Production standard 9.3.606 Parse CLI options and arguments
  * Production standard 7.3.602 Default variable value
* markit
  * Updated if statement syntax from this [] to this [[]] 
  * Production standard 1.3.614 DEBUG variable
  * Production standard 2.3.614 Log format
  * Production standard 0.3.615 --help
  * Production standard 9.3.606 Parse CLI options and arguments

### Issues
* close #84  define unique temporary directory names to support automation
* close #88  look at adding default TEST cases and update README.md

# Version  4.1.8.836
#### Release Name  4.1.8
#### Release Date  2020-10-03T15:36:36.764370-05:00 (CDT)
* Added ISSUE_TEMPLATES, security policy - SECURITY.md, CHANGELOG.md, clone.table.md, view.table.md
* Added file name(s) to beginning of git commit message, line 2, to help locate file(s) that changed 
* Added branch name to line 1, to help locate changes to file(s)
* Added default BradleyA/git-TEST-commit-automation testcases

### Features Added
* Added .github/ISSUE_TEMPLATE/feature_request.md
* Added .github/ISSUE_TEMPLATE/incident-report.md
* Added .github/ISSUE_TEMPLATE/comment.md
* Added CHANGELOG.md
* Added SECURITY.md - create repository SECURITY.md, closes #78
* Added images/clone.table.md
* Added images/construction-icon.gif
* Added images/view.table.md
* Updated README.md
* Added default BradleyA/git-TEST-commit-automation testcases for testing during code development
* markit
    * Added display_help about how to git push to more than one repository
    * Updated Production standards 5.3.559 Copyright,  2.3.578 Log format,  9.3.562 Parse CLI options and arguments, closes #76
    * Automate adding file name(s) to beginning of message, closes #75
    * Added test cases, closes #73
    * Testing changes to markit add branch to line 1, closes #69
    * Testing with branching, closes #68
    * Added repository and branch to display_help, closes #82
* find-code.sh
    * Production standards:  5.3.550 Copyright, 0.3.550 --help, 4.3.550 Documentation Language, 1.3.550 DEBUG variable, closes #74
* check-markit
    * Updated Production Standards 3.550, closes #72
    * Added test cases, closes #73
    * Correct display-usage
    * Upgraded Production standards, closes #79

### Issues
* close #68  Testing with branching
* close #69  Testing changes to markit add branch to line 1
* close #71  Review: https://github.com/commitizen/cz-cli
* close #72  Updated Production Standards 3.550
* clone #73  Added test cases

find-code.sh
#!/bin/bash
# 	find-code.sh  4.1.145.1088  2020-12-15T22:14:45.165434-06:00 (CST)  https://github.com/BradleyA/markit.git  master  uadmin  three-rpi3b.cptx86.com 4.1.144  
# 	   find-code.sh -->   testing  
# 	find-code.sh  4.1.134.1073  2020-12-06T11:47:37.157945-06:00 (CST)  https://github.com/BradleyA/markit  master  uadmin  five-rpi3b.cptx86.com 4.1.133-3-g33ebd0b  
# 	   find-code.sh -->   Production standard 1.3.614 DEBUG variable  Production standard 2.3.614 Log format (WHEN WHERE WHAT Version Line WHO UID:GID [TYPE] Message)  Production standard 0.3.615 --help  
# 	find-code.sh  4.1.108.1030  2020-11-12T13:13:54.908899-06:00 (CST)  https://github.com/BradleyA/markit  master  uadmin  five-rpi3b.cptx86.com 4.1.107 
# 	   find-code.sh -->   Production standard 9.3.606 Parse CLI options and arguments  
# 	find-code.sh  4.1.88.986  2020-10-25T22:59:40.978813-05:00 (CDT)  https://github.com/BradleyA/markit  master  uadmin  five-rpi3b.cptx86.com 4.1.87  
# 	   find-code.sh -->   Production standard 7.3.602 Default variable value  
# 	find-code.sh  4.1.22.908  2020-10-21T14:20:37.170990-05:00 (CDT)  https://github.com/BradleyA/markit  master  uadmin  five-rpi3b.cptx86.com 4.1.21  
# 	   find-code.sh -->   begin testing options  
#86# find-code.sh - Search systems from clones from repositories
###  Production standard 5.3.559 Copyright                                    # 3.559
#    Copyright (c) 2020 Bradley Allen                                                # 3.555
#    MIT License is online in the repository as a file named LICENSE"         # 3.559
###  Production standard 3.0 shellcheck
###  Production standard 1.3.614 DEBUG variable
#    Order of precedence: environment variable, default code
if [[ "${DEBUG}" == ""  ]] ; then DEBUG="0" ; fi   # 0 = debug off, 1 = debug on, 'export DEBUG=1', 'unset DEBUG' to unset environment variable (bash)
if [[ "${DEBUG}" == "2" ]] ; then set -x    ; fi   # Print trace of simple commands before they are executed
if [[ "${DEBUG}" == "3" ]] ; then set -v    ; fi   # Print shell input lines as they are read
if [[ "${DEBUG}" == "4" ]] ; then set -e    ; fi   # Exit immediately if non-zero exit status
if [[ "${DEBUG}" == "5" ]] ; then set -e -o pipefail ; fi   # Exit immediately if non-zero exit status and exit if any command in a pipeline errors
#
#
BOLD=$(tput -Txterm bold)
UNDERLINE=$(tput -Txterm sgr 0 1)  # 0.3.583
NORMAL=$(tput -Txterm sgr0)
RED=$(tput    setaf 1)
YELLOW=$(tput setaf 3)
BLUE=$(tput   setaf 4)
PURPLE=$(tput setaf 5)
CYAN=$(tput   setaf 6)
WHITE=$(tput  setaf 7)

###  Production standard 7.3.602 Default variable value
DEFAULT_CLUSTER="us-tx-cluster-1/"
DEFAULT_DATA_DIR="/usr/local/data/"
DEFAULT_SYSTEMS_FILE="SYSTEMS"

###  Production standard 8.3.541 --usage
COMMAND_NAME=$(echo "${0}" | sed 's/^.*\///')                                               # 3.541
display_usage() {
echo -e "\n${NORMAL}${COMMAND_NAME}\n   Search systems for .git repositories"
echo -e "\n${BOLD}USAGE${NORMAL}"
echo -e "   ${COMMAND_NAME} [-c <CLUSTER>] [-d <DATA_DIR>] [-s <SYSTEMS_FILE>]\n"
echo    "   ${COMMAND_NAME} [--help | -help | help | -h | h | -?]"
echo    "   ${COMMAND_NAME} [--usage | -usage | -u]"
echo    "   ${COMMAND_NAME} [--version | -version | -v]"
}

###  Production standard 0.3.615 --help
display_help() {
display_usage
#    Displaying help DESCRIPTION in English en_US.UTF-8, en.UTF-8, C.UTF-8
echo -e "\n${BOLD}DESCRIPTION${NORMAL}"
echo -e "\nThis script reads ${DEFAULT_DATA_DIR}${DEFAULT_CLUSTER}${DEFAULT_SYSTEMS_FILE} file for hosts"
echo    "to search for .git repositories in ~/.. directories."
echo -e "\nThe <DATA_DIR>/<CLUSTER>/<SYSTEMS_FILE> includes one FQDN or IP address per"
echo    "line for all hosts in the cluster.  Lines in <SYSTEMS_FILE> that begin with a"
echo    "'#' are comments.  The <SYSTEMS_FILE> is used by markit/find-code.sh,"
echo    "Linux-admin/cluster-command/cluster-command.sh, docker-TLS/copy-registry-tls.sh,"
echo    "pi-display/create-message/create-display-message.sh, and other scripts.  A"
echo    "different <SYSTEMS_FILE> can be used by setting the SYSTEMS_FILE environment"
echo    "variable or by editing this script and changing DEFAULT_SYSTEMS_FILE."
echo -e "\nThe user may receive password and/or passphrase prompts from a remote system;"
echo    "running the following may stop the prompts in your cluster."
echo -e "\t${BOLD}ssh-copy-id <TLS_USER>@<REMOTE_HOST>${NORMAL}"
echo    "or"
echo -e "\t${BOLD}ssh-copy-id <TLS_USER>@<192.168.x.x>${NORMAL}"
echo    "If that does not resolve the prompting challenge then review the man pages for"
echo    "ssh-agent and ssh-add before entering the following in a terminal window."
echo -e "\t${BOLD}eval \$(ssh-agent)${NORMAL}"
echo -e "\t${BOLD}ssh-add${NORMAL}"

###  Production standard 4.3.587 Documentation Language                                     # 3.550
#    Displaying help DESCRIPTION in French fr_CA.UTF-8, fr_FR.UTF-8, fr_CH.UTF-8
if [[ "${LANG}" == "fr_CA.UTF-8" ]] || [[ "${LANG}" == "fr_FR.UTF-8" ]] || [[ "${LANG}" == "fr_CH.UTF-8" ]] ; then
  echo -e "\n--> ${LANG}"
  echo    "<Votre aide va ici>" # Your help goes here
  echo    "Souhaitez-vous traduire la section description?" # Would you like to translate the description section?
elif ! [[ "${LANG}" == "en_US.UTF-8" ||  "${LANG}" == "en.UTF-8" || "${LANG}" == "C.UTF-8" ]] ; then  # 3.550
  new_message "${LINENO}" "${YELLOW}INFO${WHITE}" "  Your language, ${LANG}, is not supported.  Would you like to translate the description section?" 1>&2
fi

echo -e "\n${BOLD}ENVIRONMENT VARIABLES${NORMAL}"
echo    "If using the bash shell, enter; 'export DEBUG=1' on the command line to set"
echo    "the environment variable DEBUG to '1' (0 = debug off, 1 = debug on).  Use the"
echo    "command, 'unset DEBUG' to remove the exported information from the environment"
echo    "variable DEBUG.  You are on your own defining environment variables if"
echo    "you are using other shells."

###  Production standard 1.3.614 DEBUG variable
echo    "   DEBUG           (default off '0')  The DEBUG environment variable can be set"   # 1.3.550
echo    "                   to 0, 1, 2, 3, 4 or 5.  The setting '' or 0 will turn off"      # 1.3.550
echo    "                   all DEBUG messages during execution of this script.  The"       # 1.3.550
echo    "                   setting 1 will print all DEBUG messages during execution."      # 1.3.550
echo    "                   Setting 2 (set -x) will print a trace of simple commands"       # 1.3.550
echo    "                   before they are executed.  Setting 3 (set -v) will print"       # 1.3.550
echo    "                   shell input lines as they are read.  Setting 4 (set -e) will"   # 1.3.550
echo    "                   exit immediately if non-zero exit status is recieved with"      # 1.3.550
echo    "                   some exceptions.  Setting 5 (set -e -o pipefail) will do"       # 1.3.550
echo    "                   setting 4 and exit if any command in a pipeline errors.  For"   # 1.3.550
echo    "                   more information about the set options, see man bash."          # 1.3.550
#
echo    "   CLUSTER         Cluster name (default '${DEFAULT_CLUSTER}')"
echo    "   DATA_DIR        Data directory (default '${DEFAULT_DATA_DIR}')"
echo    "   SYSTEMS_FILE    Name of file that includes hosts in cluster"
echo    "                   (default '${DEFAULT_SYSTEMS_FILE}')"

echo -e "\n${BOLD}OPTIONS${NORMAL}"
echo -e "Order of precedence: CLI options, environment variable, default value.\n"                     # 0.3.595
echo    "   --help, -help, help, -h, h, -?"                                                            # 0.3.595
echo -e "\tOn-line brief reference manual\n"                                                           # 0.3.595
echo    "   --usage, -usage, -u"                                                                       # 0.3.595
echo -e "\tOn-line command usage\n"                                                                    # 0.3.595
echo    "   --version, -version, -v"                                                                   # 0.3.595
echo -e "\tOn-line command version\n"                                                                  # 0.3.595
#
echo    "   --cluster <CLUSTER>, -c <CLUSTER>, --cluster=<CLUSTER>, -c=<CLUSTER>"                      # 0.3.595
echo -e "\tCluster name (default '${DEFAULT_CLUSTER}')\n"
echo    "   --datadir <DATA_DIR>, -d <DATA_DIR>, --datadir=<DATA_DIR>, -d=<DATA_DIR>"                  # 0.3.595
echo -e "\tData directory (default '${DEFAULT_DATA_DIR}')\n"
echo    "   --systems <SYSTEMS_FILE>, -s <SYSTEMS_FILE>, --systems=<SYSTEMS_FILE>, -s=<SYSTEMS_FILE>"  # 0.3.595
echo -e "\tName of systems file (default '${DEFAULT_SYSTEMS_FILE}')"

###  Production standard 6.3.547  Architecture tree
echo -e "\n${BOLD}ARCHITECTURE TREE${NORMAL}"  # STORAGE & CERTIFICATION
echo    "/usr/local/data/                           <-- <DATA_DIR> Data directory"
echo    "└── <CLUSTER>/                             <-- <CLUSTER> Cluster name"
echo    "    └── SYSTEMS                            <-- <SYSTEMS_FILE> List of hosts in cluster"

echo -e "\n${BOLD}DOCUMENTATION${NORMAL}"
echo    "   ${UNDERLINE}https://github.com/BradleyA/markit/blob/master/README.md${NORMAL}"

echo -e "\n${BOLD}EXAMPLES${NORMAL}"
echo -e "   Search systems for .git repositories using defaults\n\t${BOLD}${COMMAND_NAME}${NORMAL}\n"  # 0.3.595
echo -e "   Search systems for .git repositories using a different <CLUSTER>.\n\t${BOLD}${COMMAND_NAME} -c australia-southeast1 ${NORMAL}\n" # 0.3.595
echo -e "   Search systems for .git repositories using a different <SYSTEMS_FILE>.\n\t${BOLD}${COMMAND_NAME} -s SYSTEMS-RASPBERRY ${NORMAL}" # 0.3.595

echo -e "\n${BOLD}SEE ALSO${NORMAL}"                                                                          #  0.3.615
echo    "   ${BOLD}check-markit${NORMAL}  Check the release version of a file"                                #  0.3.615
echo -e "\twith the remote Git repository release version"                                                    #  0.3.615
echo -e "\t(${UNDERLINE}https://github.com/BradleyA/markit/blob/master/README.md${NORMAL})\n"                 #  0.3.615
echo    "   ${BOLD}markit${NORMAL}  Mark tracked modified file(s), in your local"                             #  0.3.615
echo -e "\tGit repository and push those changes to a remote Git repository"                                  #  0.3.615
echo -e "\t(${UNDERLINE}https://github.com/BradleyA/markit/blob/master/README.md${NORMAL})"                   #  0.3.615

echo -e "\n${BOLD}AUTHOR${NORMAL}"
echo    "   ${COMMAND_NAME} was written by Bradley Allen <allen.bradley@ymail.com>"

echo -e "\n${BOLD}REPORTING BUGS${NORMAL}"                                                             # 0.3.595
echo    "   Report ${COMMAND_NAME} bugs ${UNDERLINE}https://github.com/BradleyA/markit/issues/new/choose${NORMAL}"  # 0.3.595

###  Production standard 5.3.559 Copyright
echo -e "\n${BOLD}COPYRIGHT${NORMAL}"
echo    "   Copyright (c) 2020 Bradley Allen"
echo    "   MIT License is online in the repository as a file named LICENSE"
}

#    Date and time function ISO 8601
get_date_stamp() {
  DATE_STAMP=$(date +%Y-%m-%dT%H:%M:%S.%6N%:z)
  TEMP=$(date +%Z)
  DATE_STAMP="${DATE_STAMP} (${TEMP})"
}

#    Fully qualified domain name FQDN hostname
LOCALHOST=$(hostname -f)

#    Version
#    Assumptions for the next two lines of code:  The second line in this script includes the script path & name as the second item and
#    the script version as the third item separated with space(s).  The tool I use is called 'markit'. See example line below:
#       template/template.sh  3.517.783  2019-09-13T18:20:42.144356-05:00 (CDT)  https://github.com/BradleyA/user-files.git  uadmin  one-rpi3b.cptx86.com 3.516  
SCRIPT_NAME=$(head -2 "${0}" | awk '{printf $2}')  #  Different from ${COMMAND_NAME}=$(echo "${0}" | sed 's/^.*\///'), SCRIPT_NAME = includes Git repository directory and can be used any where in script (for dev, test teams)
SCRIPT_VERSION=$(head -2 "${0}" | awk '{printf $3}')
if [[ "${SCRIPT_NAME}" == "" ]] ; then SCRIPT_NAME="${0}" ; fi
if [[ "${SCRIPT_VERSION}" == "" ]] ; then SCRIPT_VERSION="v?.?" ; fi

#    GID
GROUP_ID=$(id -g)

###  Production standard 2.3.614 Log format (WHEN WHERE WHAT Version Line WHO UID:GID [TYPE] Message)
new_message() {  #  $1="${LINENO}"  $2="DEBUG INFO ERROR WARN"  $3="message"
  get_date_stamp
  echo -e "${NORMAL}${DATE_STAMP} ${LOCALHOST} ${BOLD}${CYAN}${SCRIPT_NAME}${NORMAL}[$$] ${BOLD}${BLUE}${SCRIPT_VERSION} ${PURPLE}${1}${NORMAL} ${USER} ${UID}:${GROUP_ID} ${BOLD}[${2}]${NORMAL}  ${3}"  # 2.3.614
}

#    INFO
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "${YELLOW}INFO${WHITE}" "${BOLD}${CYAN}  Started...${NORMAL}" 1>&2 ; fi  # 1.3.614

#    Added following code because USER is not defined in crobtab jobs
if ! [[ "${USER}" == "${LOGNAME}" ]] ; then  USER=${LOGNAME} ; fi
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "DEBUG" "  Setting USER to support crobtab...  USER >${YELLOW}${USER}${WHITE}<  LOGNAME >${YELLOW}${LOGNAME}${WHITE}<" 1>&2 ; fi  #  2.3.578

#    DEBUG  #  1.3.614
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "${YELLOW}DEBUG${WHITE}" "  Name_of_command >${YELLOW}${SCRIPT_NAME}${WHITE}< Name_of_arg1 >${YELLOW}${1}${WHITE}< Name_of_arg2 >${YELLOW}${2}${WHITE}< Name_of_arg3 >${YELLOW}${3}${WHITE}<  Version of bash ${YELLOW}${BASH_VERSION}${WHITE}" 1>&2 ; fi  #  1.3.614

###  Production standard 9.3.606 Parse CLI options and arguments
while [[ "${#}" -gt 0 ]] ; do
  case "${1}" in
    --help|-help|help|-h|h|-\?)  display_help | more ; exit 0 ;;
    --usage|-usage|usage|-u)  display_usage ; exit 0  ;;
    --version|-version|version|-v)  echo "${SCRIPT_NAME} ${SCRIPT_VERSION}" ; exit 0  ;;
    -c|--cluster)  if [[ "${2}" == "" ]] ; then echo -e "\n${BOLD}    Argument for ${YELLOW}${1}${WHITE} is not found on command line.${NORMAL}\n" ; exit 1 ; fi ; CLUSTER=${2} ; shift 2  ;;  # 9.3.596
    -c=*|--cluster=*) CLUSTER="${1#*=}" ;      if [[ "${CLUSTER}" == "" ]]      ; then echo -e "\n${BOLD}    Argument for ${YELLOW}${1}${WHITE} is not found on command line.${NORMAL}\n" ; exit 1 ; fi ; shift  ;;  # 9.3.596
    -d|--datadir)  if [[ "${2}" == "" ]] ; then echo -e "\n${BOLD}    Argument for ${YELLOW}${1}${WHITE} is not found on command line.${NORMAL}\n" ; exit 1 ; fi ; DATA_DIR=${2} ; shift 2  ;;  # 9.3.596
    -d=*|--datadir=*) DATA_DIR="${1#*=}" ;     if [[ "${DATA_DIR}" == "" ]]     ; then echo -e "\n${BOLD}    Argument for ${YELLOW}${1}${WHITE} is not found on command line.${NORMAL}\n" ; exit 1 ; fi ; shift  ;;  # 9.3.596
    -s|--systems)  if [[ "${2}" == "" ]] ; then echo -e "\n${BOLD}    Argument for ${YELLOW}${1}${WHITE} is not found on command line.${NORMAL}\n" ; exit 1 ; fi ; SYSTEMS_FILE=${2} ; shift 2  ;;  # 9.3.596
    -s=*|--systems=*) SYSTEMS_FILE="${1#*=}" ; if [[ "${SYSTEMS_FILE}" == "" ]] ; then echo -e "\n${BOLD}    Argument for ${YELLOW}${1}${WHITE} is not found on command line.${NORMAL}\n" ; exit 1 ; fi ; shift  ;;  # 9.3.596
    *) echo -e "\n${BOLD}    Invalid option, ${YELLOW}${1}${WHITE}, try ${YELLOW}--usage${NORMAL}\n" ; exit 1 ; ;; # 9.3.606
  esac
done

###

###  Production standard 7.3.602 Default variable value
#    Order of precedence: CLI argument, environment variable, default code
if [[ "${CLUSTER}" == "" ]] ; then CLUSTER="${DEFAULT_CLUSTER}" ; fi
if [[ "${DATA_DIR}" == "" ]] ; then DATA_DIR="${DEFAULT_DATA_DIR}" ; fi
if [[ "${SYSTEMS_FILE}" == "" ]] ; then SYSTEMS_FILE="${DEFAULT_SYSTEMS_FILE}" ; fi
if [[ "${DEBUG}" == "1" ]] ; then new_message "${LINENO}" "${YELLOW}DEBUG${WHITE}" "  CLUSTER >${BOLD}${CYAN}${CLUSTER}${NORMAL}< DATA_DIR >${BOLD}${CYAN}${DATA_DIR}${NORMAL}< SYSTEMS_FILE >${BOLD}${CYAN}${SYSTEMS_FILE}${NORMAL}< PATH >${BOLD}${CYAN}${PATH}${NORMAL}<" 1>&2 ; fi
#
#    Check if ${DATA_DIR} directory is on system
if ! [[ -d "${DATA_DIR}" ]] ; then
  new_message "${LINENO}" "${RED}ERROR${WHITE}" "  Path to cluster data directory, ${DATA_DIR}, not found." 1>&2
  exit 1
fi
#
#    Check if ${CLUSTER} directory is on system
if ! [[ -d "${DATA_DIR}"/"${CLUSTER}" ]] ; then
  new_message "${LINENO}" "${RED}ERROR${WHITE}" "  Cluster directory name, ${CLUSTER}, not found." 1>&2
  exit 1
fi
#
#    Check if ${SYSTEMS_FILE} file contains /
if [[  "${SYSTEMS_FILE}" == *"/"* ]] ; then
  new_message "${LINENO}" "${RED}ERROR${WHITE}" "  SYSTEMS filename, ${SYSTEMS_FILE}, includes a '/'." 1>&2
  exit 1
fi
#
#    Check if ${SYSTEMS_FILE} file is on system, one FQDN or IP address per line for all hosts in cluster
if ! [[ -e "${DATA_DIR}"/"${CLUSTER}"/"${SYSTEMS_FILE}" ]] || ! [[ -s "${DATA_DIR}"/"${CLUSTER}"/"${SYSTEMS_FILE}" ]] ; then
  new_message "${LINENO}" "${YELLOW}WARN${WHITE}" "  Name of systems file, ${SYSTEMS_FILE} not found or empty.  Creating ${SYSTEMS_FILE} file and including local host.  Edit ${DATA_DIR}/${CLUSTER}/${SYSTEMS_FILE} file and add additional hosts that are in the cluster." 1>&2
  echo -e "###     List of hosts used by cluster-command.sh & create-message.sh"  > "${DATA_DIR}/${CLUSTER}/${SYSTEMS_FILE}"
  echo -e "#       One FQDN or IP address per line for all hosts in cluster" > "${DATA_DIR}/${CLUSTER}/${SYSTEMS_FILE}"
  echo -e "###" > "${DATA_DIR}/${CLUSTER}/${SYSTEMS_FILE}"
  echo    "${LOCALHOST}" > "${DATA_DIR}/${CLUSTER}/${SYSTEMS_FILE}"
fi

#    Loop through hosts in ${SYSTEMS_FILE} file
REMOTE_COMMAND="find ~/.. 2>/dev/null -type d -name '.git' -print | sed 's/^.*\.\./   \~/' | sed 's/\/\.git//'"
REMOTEHOST=$(grep -v "#" "${DATA_DIR}/${CLUSTER}/${SYSTEMS_FILE}")
for NODE in ${REMOTEHOST} ; do
  echo -e "\n${BOLD}  -->  ${CYAN}${NODE}${NORMAL}    ->${PURPLE}${REMOTE_COMMAND}${WHITE}<-"
  if [[ "${LOCALHOST}" != "${NODE}" ]] ; then
    ssh -t "${USER}"@"${NODE}" "echo -e '${BOLD}${YELLOW}\c' ; ${REMOTE_COMMAND} ; echo -e '${NORMAL}\c'"
  else
    eval "echo -e '${BOLD}${YELLOW}\c' ; ${REMOTE_COMMAND} ; echo -e '${NORMAL}\c'"
  fi
done

#
new_message "${LINENO}" "${YELLOW}INFO${WHITE}" "${BOLD}${CYAN}  Operation finished...${NORMAL}" 1>&2  # 1.3.614
###
